### ISSUE TITLE
Compartments

### ISSUE THREAD

**Issue by WilliamIX** [BuildACell/BioCRNPyler#200]
Species should be able to have an SBML compartment tag. This won't do anything when simulated with bioscrape, but will work with other SBML simulators and SubSBML.

@ayush9pandey can you help spec out what this looks like? I have a couple of thoughts....

1. Compartments can just become attributes and be automatically incorporated into Species' names. Although this isn't what SBML intended, it will make everything work with bioscrape.
2. In deterministic CRNs, the volume of a compartment doesn't matter....so for now we can ignore that. If compartments are used with the stochastic=True case, we should raise a warning.
3. When writing SBML, biocrnpyler can look over all the species and make all the compartments automatically.

In the future, it would be cool for Bioscrape to include compartments, but this would be a fairly serious new feature.

**Comment by mkapasiawala** [comment]
This would also be incredibly helpful for creating membrane models in BioCRNpyler; in particular, naming/creating species in a reproducible, scalable manner using the name of their compartments (can be easily done if there is an attribute called "compartment", as William suggested). 

At the moment, in trying to create membrane transport Mechanisms, I am having to either manually change the names of all relevant species to "{species name}_internal" and "{species name}_external". Having a simple compartment attribute would make it easier to do that. 

Another way we can do this is creating multiple Species in a single line of code. For example, let's say we want to create a Species A that is in both compartment 1 and compartment 2. Then, instead of writing 
`A_internal = Species(name="A", attributes=['compartment1'])` and `A_external = Species(name="A", attributes=['compartment2'])`,
we can write 
`A_internal, A_external = Species(name="A", compartments=['compartment1', 'compartment2'])`

I don't know how difficult this is to implement but these are just some thoughts I had.

**Comment by zoltuz** [comment]
I would prefer to add the compartment name as an instance attribute, and each reaction should do a compartment check when a reaction is created.
```python
A = Species('A', compartment='inside')
B = Species('B', compartment='outside')
if Rereaction._compartment_eq(A,B):
   pass
```
We can still rename the species when an SBML file is created for bioscrape.

**Comment by ayush9pandey** [comment]
Working on this. 

> 
> 
> I would prefer to add the compartment name as an instance attribute, and each reaction should do a compartment check when a reaction is created.
I like this, it's how I thought of implementing compartments. I have a first cut of this in the compartments branch. 
> 
> ```python
> A = Species('A', compartment='inside')
> B = Species('B', compartment='outside')
> if Rereaction._compartment_eq(A,B):
>    pass
> ```
> 
Actually, this check is not needed because when creating membrane models you need reactions between two different compartments (that is for transport reactions). Since this is the main purpose why someone would create species with compartments, we can't have this check. Are there any other checks needed?


> We can still rename the species when an SBML file is created for bioscrape.
There is no issue with writing SBML for bioscrape with regards to compartments. All species have to have unique IDs, that is what bioscrape looks for so all biocrnpyler needs to do is ensure that all species have unique IDs.

**Comment by ayush9pandey** [comment]
> 
> 
> This would also be incredibly helpful for creating membrane models in BioCRNpyler; in particular, naming/creating species in a reproducible, scalable manner using the name of their compartments (can be easily done if there is an attribute called "compartment", as William suggested).
> 
> At the moment, in trying to create membrane transport Mechanisms, I am having to either manually change the names of all relevant species to "{species name}_internal" and "{species name}_external". Having a simple compartment attribute would make it easier to do that.
> 
> Another way we can do this is creating multiple Species in a single line of code. For example, let's say we want to create a Species A that is in both compartment 1 and compartment 2. Then, instead of writing
> `A_internal = Species(name="A", attributes=['compartment1'])` and `A_external = Species(name="A", attributes=['compartment2'])`,
> we can write
> `A_internal, A_external = Species(name="A", compartments=['compartment1', 'compartment2'])`
> 
> I don't know how difficult this is to implement but these are just some thoughts I had.

Thanks for example use cases @mkapasiawala. just as a note, you can use Sub-SBML until biocrnpyler compartments are ready for membrane models. But hopefully soon you can use biocrnpyler directly.

**Comment by ayush9pandey** [comment]
> 
> 
> Species should be able to have an SBML compartment tag. This won't do anything when simulated with bioscrape, but will work with other SBML simulators and SubSBML.
> 
> @ayush9pandey can you help spec out what this looks like? I have a couple of thoughts....
> 
>     1. Compartments can just become attributes and be automatically incorporated into Species' names. Although this isn't what SBML intended, it will make everything work with bioscrape.
> 
>     2. In deterministic CRNs, the volume of a compartment doesn't matter....so for now we can ignore that. If compartments are used with the stochastic=True case, we should raise a warning.

Do we need this warning from biocrnpyler though? It seems like a simulator job to make sure that if your model has compartment volume and you are trying to run stochastic simulation using that simulator => the simulator warns you about it.


> 
>     3. When writing SBML, biocrnpyler can look over all the species and make all the compartments automatically.
> 
> 
> In the future, it would be cool for Bioscrape to include compartments, but this would be a fairly serious new feature.

**Comment by ayush9pandey** [comment]
What do we want to do for other properties of a compartment for example:

`Species("A", compartment = "new_compartment")` only specifies the name of the compartment. But, in SBML, a compartment can have an integer number of "spatialDimensions" and a float of "volume". 

Do we want biocrnpyler to be able to set these as well or just set a compartment for a species?

**Comment by WilliamIX** [comment]
I think we will want to do this eventually, but it isn't a high priority in terms of the paper (maybe something we can include in a follow up release). Ultimately, having compartments means we want multiple volumes, and potentially different spatial dimensions. It would be very cool to include these features in an optimized way in Bioscrape.

**Comment by ayush9pandey** [comment]
Yeah, the real challenge is when doing all this in bioscrape. For biocrnpyler, this isn't too hard to implement.

**Comment by WilliamIX** [comment]
Allow volume and dimension to be Parameters

**Comment by ayush9pandey** [comment]
Fixed with #211 .
